### PR TITLE
Implement oauth with guest token retrieval

### PR DESCRIPTION
**Pull request by mjurkus**
Basic oauth implementation. No retries, no expired token handling, no nothing at this point.

But with this it is already possible to develop new API. Will add error handling/token refresh in separate PRs.

### CODE REVIEW

**Comment by mjurkus** [comment]
cc @vycius

File: app/build.gradle [hunk 1]
```diff
@@ -108,6 +108,7 @@ dependencies {
 
     implementation Libs.rx_android
     implementation Libs.rx_java
+    implementation Libs.rx_java_2
```
Comment (mjurkus) [inline app/build.gradle:111]: Need to migrate all Rx to Rx2

File: app/src/main/java/lt/vilnius/tvarkau/api/ApiHeadersInterceptor.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+        if (apiToken.isSet() && !containsTokenHeader) {
+            applyToken(requestBuilder)
+        }
+    }
+
+    public fun applyToken(requestBuilder: Request.Builder) {
+        val tokenValue = apiToken.get()
+        requestBuilder.addHeader(HTTP_HEADER_OAUTH, formatTokenForHeader(tokenValue.accessToken))
+        println("header: " + formatTokenForHeader(tokenValue.accessToken))
```
Comment (vycius) [inline app/src/main/java/lt/vilnius/tvarkau/api/ApiHeadersInterceptor.kt]: No need for print

File: app/src/main/java/lt/vilnius/tvarkau/api/ApiHeadersInterceptor.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    private fun appendToken(chain: Interceptor.Chain, requestBuilder: Request.Builder) {
+        val containsTokenHeader = chain.request().headers().names().contains(HTTP_HEADER_OAUTH)
+
+        if (apiToken.isSet() && !containsTokenHeader) {
+            applyToken(requestBuilder)
+        }
+    }
+
+    public fun applyToken(requestBuilder: Request.Builder) {
```
Comment (vycius) [inline app/src/main/java/lt/vilnius/tvarkau/api/ApiHeadersInterceptor.kt]: Public by default

File: app/src/main/java/lt/vilnius/tvarkau/dagger/module/DataModule.kt [hunk 4]
```diff
@@ -47,4 +50,34 @@ class DataModule {
...
         )
     }
-}
\ No newline at end of file
+
+    @Provides
+    @Singleton
+    @RawOkHttpClient
+    fun provideRawHttpClient(
+            cache: okhttp3.Cache,
```
Comment (vycius) [inline app/src/main/java/lt/vilnius/tvarkau/dagger/module/DataModule.kt]: Be constant either it is `Cache` in all places including `provideNetworkCache` or `okhttp3.Cache` all places.

File: app/src/main/java/lt/vilnius/tvarkau/dagger/module/SerializationModule.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+@Module
+class SerializationModule {
+
+    @Provides
+    fun provideGson(): Gson {
+        return GsonBuilder()
+                .registerTypeAdapter(LocalDateTime::class.java, LegacyApiModule.LocalDateTimeSerializer())
+                .serializeNulls()
+                .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
```
Comment (vycius) [inline app/src/main/java/lt/vilnius/tvarkau/dagger/module/SerializationModule.kt:20]: I would avoid setting `FieldNamingPolicy` for `SerializationModule` which is should be used all over the project and not just for API calls. In this case i would be more explicitly by adding `@SerializedName` in places where it's needed.

File: app/src/test/java/lt/vilnius/tvarkau/auth/SessionTokenImplTest.kt [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        val result = fixture.refreshGuestToken().test()
+
+        result.assertError { true }
+    }
+
+    private fun mockOAuthResponse(
+            success: Boolean,
+            token: String = "someOtherToken"
+    ) {
+        val gson = GsonBuilder().create()
```
Comment (vycius) [inline app/src/test/java/lt/vilnius/tvarkau/auth/SessionTokenImplTest.kt]: Doesn't seem to be used

**Review by vycius — APPROVED** [review]
Looks good. Just few small comments and considerations.

File: app/src/main/java/lt/vilnius/tvarkau/dagger/module/SerializationModule.kt (hunk 5 above)
Comment (mjurkus) [reply app/src/main/java/lt/vilnius/tvarkau/dagger/module/SerializationModule.kt:20]: Intention is to use this naming policy all overt he project. All models in new API are described this way.

**Review by vycius — APPROVED** [review]

**Pull request merged**